utilities/env_librados.cc

```diff
@@ -0,0 +1,1486 @@
+   */
+  bool IsSyncThreadSafe() const {
+    // TODO:
+    // if true,  need a mutex around the position and/or bufferlist members
+    return true;
+  }
+
+  /**
+   * @brief Indicates the upper layers if the current WritableFile implementation uses direct IO.
+   * @details [long description]
+   * @return [description]
+   */
+  bool UseDirectIO() const {
+    return false;
+  }
+
+  /**
+   * @brief Get file size
+   * @details
+   *  This API will use cached file_size. Correctness is only garuanteed if file is modified by one thread.
+   * @return [description]
+   */
+  uint64_t GetFileSize() {
+    LOG_DEBUG("%lld|%lld\n", (long long)_buffer_size, (long long)_file_size);
+    return _file_size + _buffer_size;
+  }
+
+  /**
+   * @brief For documentation, refer to RandomAccessFile::GetUniqueId()
+   * @details [long description]
+   *
+   * @param id [description]
+   * @param max_size [description]
+   *
+   * @return [description]
+   */
+  size_t GetUniqueId(char* id, size_t max_size) const {
+    // All fid has the same db_id prefix, so we need to ignore db_id prefix
+    size_t s = min(max_size, FID_LEN - DB_ID_LEN - 1);
+    strncpy(id, _fid.c_str() + DB_ID_LEN + 1, s);
+    return s;
+  }
+
+  /**
+   * @brief noop
+   * @details [long description]
+   *
+   * @param offset [description]
+   * @param length [description]
+   *
+   * @return [description]
+   */
+  rocksdb::Status InvalidateCache(size_t offset, size_t length) {
+    return rocksdb::Status::OK();
+  }
+
+  using rocksdb::WritableFile::RangeSync;
+  /**
+   * @brief noop
+   * @details [long description]
+   *
+   * @param offset [description]
+   * @param nbytes [description]
+   *
+   * @return [description]
+   */
+  rocksdb::Status RangeSync(off_t offset, off_t nbytes) {
+    return rocksdb::Status::OK();
+  }
+
+protected:
+  using rocksdb::WritableFile::Allocate;
+  /**
+   * @brief noop
+   * @details [long description]
+   *
+   * @param offset [description]
+   * @param len [description]
+   *
+   * @return [description]
+   */
+  rocksdb::Status Allocate(off_t offset, off_t len) {
+    return rocksdb::Status::OK();
+  }
+};
+
+
+// Directory object represents collection of files and implements
+// filesystem operations that can be executed on directories.
+class LibradosDirectory : public rocksdb::Directory {
+  librados::IoCtx * _io_ctx;
+  std::string _fid;
+public:
+  explicit LibradosDirectory(librados::IoCtx * io_ctx, std::string fid):
+    _io_ctx(io_ctx), _fid(fid) {}
+
+  // Fsync directory. Can be called concurrently from multiple threads.
+  rocksdb::Status Fsync() {
+    return rocksdb::Status::OK();
+  }
+};
+
+// Identifies a locked file.
+// This is exclusive lock and can't nested lock by same thread
+class LibradosFileLock : public rocksdb::FileLock {
+  librados::IoCtx * _io_ctx;
+  const std::string _obj_name;
+  const std::string _lock_name;
+  const std::string _cookie;
+  int lock_state;
```

Comment:
I don't really understand what the FileLock is used for. 

If it is used to coordinate access between different threads in rocksdb (compaction vs writes, say), then we can just use some in-memory structures to track the locks.

If it is used to prevent another instances of rocksdb from opening up the database at the same time that we do, then the rados locking is important.. but it still seems weird, since we'd probably want to lock the entire database, not just individual files.

I'm pretty sure its the first case...